examples/asset/remote_asset_loader.rs

```diff
@@ -0,0 +1,167 @@
+//! This example also showcase how to use [`AssetWriter`], thus will won't work on
+//! wasm or Android, since those platforms doesn't have [`AssetWriter`].
+
+use bevy::{
+    asset::io::{
+        AssetReader, AssetReaderError, AssetSource, AssetSourceId, AssetWriter, PathStream, Reader,
+    },
+    prelude::*,
+    utils::BoxedFuture,
+};
+use bevy_internal::asset::io::AssetWriterError;
+use futures_lite::AsyncRead;
+use std::{path::Path, sync::Arc};
+
+/// A remote asset loader implementation that will try to read asset local and, if not found,
+/// download it remotely from a CDN.
+struct RemoteAssetLoader {
+    cdn_prefix: String,
+    reader: Box<dyn AssetReader>,
+    writer: Box<dyn AssetWriter>,
+}
+
+impl RemoteAssetLoader {
+    fn new() -> Self {
+        // Create the root directory if it doesn't exists on local disk.
+        let create_root_dir = true;
+        let writer = AssetSource::get_default_writer("assets".to_string())(create_root_dir)
+            .expect("Current platform doesn't support asset writing.");
+        let reader = AssetSource::get_default_reader("assets".to_string())();
+
+        // This is the CDN which will be downloaded the font when missing localy.
+        let cdn_prefix =
+            "https://github.com/google/fonts/raw/main/ofl/protestrevolution/".to_string();
+
+        RemoteAssetLoader {
+            reader,
+            writer,
+            cdn_prefix,
+        }
+    }
+
+    /// Remote download the asset from CDN.
+    async fn read_remote<'a>(
+        &'a self,
+        path: &'a Path,
+    ) -> Result<Box<dyn AsyncRead + Send + Sync + Unpin + 'a>, AssetReaderError> {
+        // Only the asset name is needed, not the folder.
+        let asset_name = path.file_name().unwrap().to_str().unwrap();
+        let mut url = self.cdn_prefix.clone();
+        url.push_str(asset_name);
+
+        // A simple GET request is used, but you could set custom headers, auth and so on.
+        let request = ehttp::Request::get(url);
+
+        let body = match ehttp::fetch_async(request).await {
+            Ok(response) => {
+                // Since this is an example, only check for 200 status, but in a real world use
+                // it would be wise to check for others 2xx or 3xx status.
+                if response.status != 200 {
+                    return Err(AssetReaderError::HttpError(response.status));
+                }
+
+                response.bytes
+            }
+            Err(error) => {
+                warn!("Failed to read remote asset: {error}");
+                return Err(AssetReaderError::HttpError(500));
+            }
+        };
+
+        // Try to save the downloaded asset on disk.
+        if let Err(AssetWriterError::Io(error)) = self.writer.write_bytes(path, &body).await {
+            return Err(AssetReaderError::Io(Arc::new(error)));
+        }
+
+        // Try to read again the asset, since its saved on disk now.
+        self.reader.read(path).await
```

Comment:
would it work on platforms without an `AssetWriter` if it returns just the body in a wrapper implementing `AsyncReader` on those platforms?

Reply:
I think so, since `AsynReader` is implemented for any `&[u8]`, but for wasm and Android the "caching" wont work, not until we have `AssetWriter` working there.

This example will automatically work whenever a default `AssetWriter` is implemented for wasm or Android.

Reply:
I was able to do that, but it on my local tests, I kept getting errors, due to CORS policy and `ehttp` has no way to set `no-cors` mode atm, since it is fixed on code (https://github.com/emilk/ehttp/blob/master/ehttp/src/web.rs#L45C5-L45C43).

I suspect it will work on bevy website, but I don't feel confident sending a wasm example without testing it locally.